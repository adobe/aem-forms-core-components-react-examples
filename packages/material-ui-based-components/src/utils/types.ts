import { FieldJson, FieldsetJson, State } from '@aemforms/af-core';
import { Handlers, WithViewState } from '@aemforms/af-react-renderer';

export type FieldViewState = WithViewState<FieldJson>;

export type PROPS = State<FieldJson & Handlers & {
  isError?: boolean,
  layout?: {
    [key: string]: any
  }
  text?:string,
  richText?:boolean
}>;

export type PROPS_PANEL = State<FieldsetJson> & { handlers: Handlers }
