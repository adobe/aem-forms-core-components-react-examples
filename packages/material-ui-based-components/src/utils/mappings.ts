import Form from '../components/Form';
import Panel from '../components/Panel';
import TextField from '../components/TextField';
import Date from '../components/Date';
import PlainText from '../components/PlainText';
import Title from '../components/Title';
import Image from '../components/Image';
import Rating from '../components/Rating';
import Slider from '../components/Slider';
import TabsVerticalComp from '../components/TabsVerticalComp';
import TabsHorizontalComp from '../components/TabsHorizontalComp';

export default {
  form: Form,
  panel: Panel,
  'text-input': TextField,
  'date-input': Date,
  'text-input': TextField,
  'plain-text': PlainText,
  'custom:title': Title,
  image: Image,
  'text-input': TextField,
  'custom:rating':Rating,
  'custom:slider':Slider
  'text-input': TextField,
  'custom:vertical-tab': TabsVerticalComp,
  'custom:horizontal-tab': TabsHorizontalComp
};