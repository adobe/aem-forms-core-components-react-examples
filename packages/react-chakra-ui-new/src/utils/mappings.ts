// *******************************************************************************
//  * Copyright 2023 Adobe
//  *
//  * Licensed under the Apache License, Version 2.0 (the “License”);
//  * you may not use this file except in compliance with the License.
//  * You may obtain a copy of the License at
//  *
//  *     http://www.apache.org/licenses/LICENSE-2.0
//  *
//  * Unless required by applicable law or agreed to in writing, software
//  * distributed under the License is distributed on an “AS IS” BASIS,
//  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  * See the License for the specific language governing permissions and
//  * limitations under the License.
//  ******************************************************************************

import Button from '../components/Button';
import CheckBox from '../components/CheckBox';
import CheckBoxGroup from '../components/CheckBoxGroup';
import DropDown from '../components/DropDown';
import Number from '../components/Number';
import RadioGroup from '../components/RadioButton';
import TextField from '../components/TextField';
import Panel from '../components/Panel';
import Wizard from '../components/Wizard';
import Accordion from '../components/Accordion';
import HorizontalTab from '../components/HorizontalTabs';
import VerticalTab from '../components/VerticalTabs';
import Form from '../components/Form';
import Image from '../components/Image';
import PlainText from '../components/PlainText';

const mappings = {
  'text-input': TextField,
  'password-input': TextField,
  'multiline-input': TextField,
  panel: Panel,
  'checkbox-group': CheckBoxGroup,
  'radio-group': RadioGroup,
  'drop-down': DropDown,
  'number-input': Number,
  button: Button,
  checkbox: CheckBox,
  image: Image,
  'core/fd/components/form/wizard/v1/wizard': Wizard,
  'core/fd/components/form/accordion/v1/accordion': Accordion,
  'core/fd/components/form/tabsontop/v1/tabsontop': HorizontalTab,
  'core/fd/components/form/verticaltabs/v1/verticaltabs': VerticalTab,
  'plain-text': PlainText,
  'core/fd/components/form/panelcontainer/v1/panelcontainer': Panel,
  form: Form
};

export default mappings;