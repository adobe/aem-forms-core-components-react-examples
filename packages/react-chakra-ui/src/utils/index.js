import React from 'react';
import { FormContext } from '@aemforms/af-react-renderer';
import { createFormInstance } from '@aemforms/af-core';
import { ChakraProvider } from '@chakra-ui/react'
import { IntlProvider } from 'react-intl';
import { render } from '@testing-library/react';

export const DEFAULT_ERROR_MESSAGE = 'There is an error in the field';

export const createForm = (field) => {
  const formJson = {
    items: [field]
  };
  return createFormInstance(formJson);
};

export const Provider = (form, mappings, locale, dictionaries) => props => {
  const c = {
    form,
    mappings,
    modelId: '$form',
    refMap: {}
  };
  const { children } = props;
  return (
    <ChakraProvider>
      <IntlProvider locale={locale} messages={dictionaries}>
        <FormContext.Provider value={c}>{children}</FormContext.Provider>
      </IntlProvider>
    </ChakraProvider>
  );
};

export const renderComponent = function (Component) {
  const test = (field, operation) => {
    const form = createForm(field);
    if (operation) {
      operation(form, form.items[0]);
    }
    const e = form.items[0].getState();
    let component = <Component {...e} />;
    const wrapper = Provider(form);
    const renderResponse = render(component, { wrapper });
    return {
      renderResponse,
      form,
      element: form?.items[0],
      component
    };
  };
  return test;
};
