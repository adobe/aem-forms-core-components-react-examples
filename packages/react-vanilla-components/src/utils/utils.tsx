export const formatBytes = (bytes: number, decimals = 0) => {
    if (!+bytes) {
      return '0 Bytes';
    }
    const k = 1024;
    const dm = decimals < 0 ? 0 : decimals;
    const sizes = ['bytes', 'kb', 'mb', 'gb', 'tb', 'pb'];
    const i = Math.floor(Math.log(bytes) / Math.log(k));

    return `${parseFloat((bytes / Math.pow(k, i)).toFixed(dm))} ${sizes[i]}`;
  };

  export const syncAriaDescribedBy = (id: string, tooltip?: string, description?: string, error?: string) => {
    const descriptions = [];

    if (description) {
        descriptions.push(`${id}__longdescription`);
    }
    if (tooltip) {
        descriptions.push(`${id}__shortdescription`);
    }
    if (error) {
        descriptions.push(`${id}__errormessage`);
    }

    return descriptions.join(' ');
};