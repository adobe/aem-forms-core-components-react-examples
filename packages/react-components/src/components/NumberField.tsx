import React from 'react';
import {State, FieldJson} from '@aemforms/af-core';
import { IHandler } from '../utils/type';
import withRuleEngine from '../utils/HOC';

const NumberField = (props: State<FieldJson> & IHandler) => {
   const { id, label, value, required, placeholder, name, maximum, minimum, description, errorMessage, valid, readOnly } = props;
   const validateState = valid === false ? 'invalid' : 'valid';
   const error = validateState === 'invalid';
   const changeHandler = (event: React.ChangeEvent<HTMLInputElement>) => {
      props.dispatchChange(event.target.value);
   };

   const handleFocus = () => {
      props.dispatchFocus();
   };

   return (
      <div className="cmp-adaptiveform-numberinput">
        <div>{label?.visible && <label id={`${id}-label`} htmlFor={id}>{label?.value}</label>}</div>
        <input type='number' className="cmp-adaptiveform-numberinput__widget" value={value} onChange={changeHandler} required={required} placeholder={placeholder} name={name} min={minimum} max={maximum} readOnly={readOnly} onFocus={handleFocus} />
        {error && <div  id={`${id}-errorMessage`} className="cmp-adaptiveform-numberinput__errormessage">{errorMessage}</div>}
           {description && !error && <div className="cmp-adaptiveform-numberinput__longdescription" id={`${id}-longDescription`}>{description}</div>}
      </div>
   ); 
};

export default withRuleEngine(NumberField);