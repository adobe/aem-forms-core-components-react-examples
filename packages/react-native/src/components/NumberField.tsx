import { State, FieldJson } from '@aemforms/af-core';
import { useRenderer } from '@aemforms/af-react-renderer';
import { Input } from 'native-base';
import { combineConvertors, baseConvertor, numberFieldConvertor } from '../utils/mappers';
import InputField from '../shared/InputField';

const Comp = InputField(Input);
const NumberComponent = function (props: State<FieldJson>) {
  const renderedComponent = useRenderer(props, Comp, combineConvertors(baseConvertor, numberFieldConvertor));
  return renderedComponent;
};

export default NumberComponent;