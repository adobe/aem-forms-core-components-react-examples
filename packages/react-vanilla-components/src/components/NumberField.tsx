// *******************************************************************************
//  * Copyright 2022 Adobe
//  *
//  * Licensed under the Apache License, Version 2.0 (the “License”);
//  * you may not use this file except in compliance with the License.
//  * You may obtain a copy of the License at
//  *
//  *     http://www.apache.org/licenses/LICENSE-2.0
//  *
//  * Unless required by applicable law or agreed to in writing, software
//  * distributed under the License is distributed on an “AS IS” BASIS,
//  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  * See the License for the specific language governing permissions and
//  * limitations under the License.

//  * The BEM markup is as per the AEM core form components guidelines.
//  * LINK-https://github.com/adobe/aem-core-forms-components/blob/master/ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/numberinput.html
//  ******************************************************************************

import React, { useState } from 'react';
import { withRuleEngine } from '../utils/withRuleEngine';
import { PROPS } from '../utils/type';

const NumberField = (props: PROPS) => {
  const { id, label, value, required, placeholder, name, maximum, minimum, description, errorMessage, isError, readOnly, visible, enabled } = props;

  const [shortDescription, setShortDescription] = useState(true);
  const [longDescription, setLongtDescription] = useState(false);

  const handleClick = (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
    event.preventDefault();
    setShortDescription(!shortDescription);
    setLongtDescription(!longDescription);
  };

  const changeHandler = (event: React.ChangeEvent<HTMLInputElement>) => {
    props.dispatchChange(event.target.value);
  };

  const handleBlur = (event: React.ChangeEvent<HTMLInputElement>) => {
    props.dispatchBlur(event.target.value);
  };

  const handleFocus = () => {
    props.dispatchFocus();
  };

  return (
    <div className="cmp-adaptiveform-numberinput" data-cmp-is="adaptiveFormNumberInput" data-cmp-visible={visible} data-cmp-enabled={enabled}>
      <div>{label?.visible && <label htmlFor={`${id}-widget`}>{label?.value}</label>}</div>
      <div>{description && <button className="cmp-adaptiveform-numberinput__questionmark" aria-label='Toggle Button' onClick={handleClick}></button>}</div>
      <input id={`${id}-widget`} type='number' className={value ? 'cmp-adaptiveform-numberinput__widget cmp-adaptiveform-numberinput__widget--filled' : 'cmp-adaptiveform-numberinput__widget cmp-adaptiveform-numberinput__widget--empty'} value={value} onChange={changeHandler} required={required} placeholder={placeholder} name={name} min={minimum} max={maximum} readOnly={readOnly} onFocus={handleFocus} onBlur={handleBlur} />
      {shortDescription && props?.tooltip && <div title='Help Text' data-cmp-visible={shortDescription} className='cmp-adaptiveform-numberinput__shortdescription'>{props?.tooltip}</div>}
      <div aria-live="polite">
        {longDescription && description && !errorMessage ? <div title='Help Text' data-cmp-visible={longDescription} className="cmp-adaptiveform-numberinput__longdescription">{description}</div> : null}
      </div>
      {isError ? <div className="cmp-adaptiveform-numberinput__errormessage">{errorMessage}</div> : null}
    </div>
  );
};

export default withRuleEngine(NumberField);