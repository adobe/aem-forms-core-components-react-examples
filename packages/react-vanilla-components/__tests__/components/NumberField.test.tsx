import NumberField from '../../src/components/NumberField';
import userEvent from "@testing-library/user-event";
import { renderComponent, DEFAULT_ERROR_MESSAGE } from '../utils';

const field = {
  name: 'number',
  label: {
    value: 'number field',
  },
  fieldType: 'number-input',
  placeholder: 'enter number field',
  visible: true,
  required: true,
};
const helper = renderComponent(NumberField);

describe('Number Field', () => {

  test('value entered by user in text field is set in model', async () => {
    const f = {
      ...field,
    };
    const { renderResponse, element } = await helper(f);
    const input = await renderResponse.findByPlaceholderText(f.placeholder);
    const inputVal = '1992';
    userEvent.type(input, inputVal);
    const state = element.getState();
    expect(state.value).toEqual(1992);
  });

  test('In case of both tooltip and description, tooltip should be visible and onclick of toggle button, description should be visible', async () => {
    const f = {
      ...field,
      tooltip: 'Short Description',
      description: 'Mandatory'
    };
    const { renderResponse } = await helper(f);
    expect(renderResponse.getByText('Short Description')).not.toBeNull();
    const button = renderResponse.container.getElementsByClassName('cmp-adaptiveform-numberinput__questionmark');
    userEvent.click(button[0]);
    expect(renderResponse.getByText('Mandatory')).not.toBeNull();
  });

  test('label is null if title is marked as hidden in the field', async () => {
    const f = {
      ...field,
      label: {
        ...field.label,
        visible: false
      }
    };
    const { renderResponse } = helper(f);
    expect(renderResponse.queryByText(f.label.value)).toBeNull();
  });

  test('labels and inputs are linked with for and id attribute', async () => {
    let { renderResponse } = await helper(field);
    const input = await renderResponse.findByPlaceholderText(field.placeholder);
    const label = await renderResponse.queryByText(field.label.value)
    expect(input?.getAttribute('id')).toEqual(label?.getAttribute('for'));
  });

  test('html in the label should be handled for non rich text', async () => {
    const f = {
      ...field,
      label: {
        value: '<p>title inside p tags</p>',
        richText: true,
        visible: true
      }
    }
    let { renderResponse } = await helper(f);
    expect(renderResponse.container.innerHTML).toContain('<p>title inside p tags</p>');
  });
});