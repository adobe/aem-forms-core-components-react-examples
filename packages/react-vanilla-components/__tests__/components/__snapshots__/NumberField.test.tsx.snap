// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Number Field Snapshot 1`] = `
<div>
  <div
    class="cmp-adaptiveform-numberinput cmp-adaptiveform-numberinput--empty "
    data-cmp-enabled="true"
    data-cmp-is="adaptiveFormNumberInput"
    data-cmp-visible="true"
  >
    <div
      class="cmp-adaptiveform-numberinput__label-container"
    >
      <label
        class="cmp-adaptiveform-numberinput__label"
        for="o2pjkfs8EnX-widget"
      >
        number field
      </label>
    </div>
    <input
      class="cmp-adaptiveform-numberinput__widget"
      id="o2pjkfs8EnX-widget"
      name="number"
      placeholder="enter number field"
      required=""
      type="number"
      value=""
    />
    <div
      aria-live="polite"
    />
  </div>
</div>
`;
